src/main/java/alfio/config/WebSecurityConfig.java

```diff
@@ -339,11 +413,244 @@ protected void configure(HttpSecurity http) throws Exception {
+                }
+
+                List<String> alfioOrganizationAuthorizationsRaw = idTokenClaims.get(openIdAuthenticationManager.getAlfioGroupsNameParameter()).asList(String.class);
+                Map<String, Set<String>> alfioOrganizationAuthorizations = new HashMap<>();
+
+                for (String alfioOrgAuth : alfioOrganizationAuthorizationsRaw) {
+                    String[] orgRole = alfioOrgAuth.split("/");
+                    String organization = orgRole[1];
+                    String role = orgRole[2];
+
+                    if (alfioOrganizationAuthorizations.containsKey(organization)) {
+                        alfioOrganizationAuthorizations.get(organization).add(role);
+                        continue;
+                    }
+                    alfioOrganizationAuthorizations.put(organization, Set.of(role));
+                }
+                Set<Role> alfioRoles = extractAlfioRoles(alfioOrganizationAuthorizations);
+                return new OpenIdAlfioUser(idToken, subject, email, false, alfioRoles, alfioOrganizationAuthorizations);
+            }
+
+            private Map<String, Object> retrieveClaims(String claimsUrl, String code) throws IOException, InterruptedException {
+                HttpClient client = HttpClient.newHttpClient();
+
+                HttpRequest request = HttpRequest.newBuilder()
+                    .uri(URI.create(claimsUrl))
+                    .header("Content-Type", openIdAuthenticationManager.getContentType())
+                    .POST(HttpRequest.BodyPublishers.ofString(openIdAuthenticationManager.buildRetrieveClaimsUrlBody(code)))
+                    .build();
+
+                HttpResponse<String> response = client.send(request,
+                    HttpResponse.BodyHandlers.ofString());
+
+                Map<String, Object> map = new ObjectMapper().readValue(response.body(), Map.class);
+                return map;
+            }
+
+            private void updateRoles(Set<Role> roles, String username) {
+                authorityRepository.revokeAll(username);
+                roles.forEach(role -> authorityRepository.create(username, role.getRoleName()));
+            }
+
+            @Override
+            public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws AuthenticationException, IOException, ServletException {
+                String code = request.getParameter(openIdAuthenticationManager.getCodeNameParameter());
+                if (code == null) {
+                    throw new IllegalArgumentException("authorization code cannot be null");
+                }
+
+                String claimsUrl = openIdAuthenticationManager.buildClaimsRetrieverUrl();
+                Map<String, Object> claims;
+                try {
+                    claims = retrieveClaims(claimsUrl, code);
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
```

Comment:
Move the "try / catch {throw}" inside the retrieveClaims method so the flow is more linear, log the exception, don't use `e.printStackTrace();`

Reply:
refactored